### ISSUE TITLE
Stella doesn't load ROM with special characters in name

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#751]
Stella seems to have problems with certain special chars in ROM names (File not found/readable). 

E.g. the 'ż' in the attached file is not liked. 
[Gruniożerca.zip](https://github.com/stella-emu/stella/files/5764486/Gruniozerca.zip)

**Comment by sa666666** [comment]
Works for me in Linux at least.  The filename is not fully displayed in the launcher, since the character set doesn't contain that character.  But Stella is still able to load it.  So if this a Windows-only thing, I suspect it's somewhere in `src/windows/FSNodeWindows`.

**Comment by thrust26** [comment]
Thanks, I was already wondering why James hadn't noticed that.

**Comment by sa666666** [comment]
I will try to test on some other systems later today.

**Comment by thrust26** [comment]
In the project settings "Multi-Byte Character Set" (MBCS) is enabled, not Unicode (which doesn't compile). 

Could that be the problem?

**Comment by sa666666** [comment]
It's possible, but I've never really played with it.  What happens when you force the Unicode `#ifdef` to be compiled?   Is this the part that won't compile?

**Comment by thrust26** [comment]
No, lots of stuff in `FilesystemNodeWINDOWS` is breaking, especially `string.c_str()` isn't accepted anymore (LPCWSTR is expected). I will try to fix that...

**Comment by sa666666** [comment]
Yeah, all that code is very ASCII-centric.  I don't know that it's ever been tested with other character sets.

**Comment by sa666666** [comment]
Incidentally, this is why I'd hoped to move to the `std::filesystem` stuff in C++17, to eliminate a big chunk of this platform-specific code.  But it seems not all compilers fully support it yet.

**Comment by thrust26** [comment]
Grrr, got the code compiled with Unicode, but it still doesn't load the ROM. The 'ż' is converted to a simple 'z'.

**Comment by sa666666** [comment]
I think we need to drill down to see exactly what is failing here.  Is it on our side, or is it happening somewhere in one of the functions that Windows is using.

Sorry I can't help more with this; currently recording lectures for next semester.  Maybe we have to bump this one to post 6.5.

**Comment by thrust26** [comment]
No hurry, this can wait. I think Windows does it correct, but Stella partially converts special chars. Either we keep the special chars completely or we ignore them completely. Currently it seems like a mix of it.

**Comment by thrust26** [comment]
I wonder what Linux is doing. Since we are using strings, it cannot store the special chars. Which means it must match 'ż' with 'z' somehow.

**Comment by thrust26** [comment]
Meanwhile I found a workaround. In `addFile()` we can use the 8.3 filename for `_path`  and the long filename for `_displayName`. 

@sa666666 What do you think?

**Comment by thrust26** [comment]
@sa666666 This fix allows loading files with specials chars.

**BUT** when saving a file, the `_path` name is used  if not entry in our DB exists for the ROM. This causes ugly names like 'Grunio~01.png'. We would have to override `getNameWithExt` for Windows and use the `_displayName` here. But that feels a bit like a hack to me.

**Comment by sa666666** [comment]
I have some ideas, but no time right now.  I will come back to this for the next release.

**Comment by sa666666** [comment]
I plan to work on this one next.  Ideally, there's a way to turn all special characters into their ASCII equivalent.  That's the best we can do, since Stella is essentially ASCII-only; the fonts, font renderer, etc all depend on this.

**Comment by thrust26** [comment]
Might become complicated. How about replacing them all e.g. with '_'?

**Comment by sa666666** [comment]
I'm not looking to hand-code a parser for this.  I'm hoping to find something in C++11 or above `locale` stuff to handle it.  Basically, any umlauts, graves, etc would be removed and converted to the 'base' ASCII character.  I don't know if this is even possible.  Barring that, converting to '_' might work too.

**Comment by thrust26** [comment]
Remember that it is not so much of a display problem, but a file access problem.

**Comment by sa666666** [comment]
Right.  For me in Linux, it is strictly a display issue.  The file load part is fine.  So this could get complicated.

**Comment by thrust26** [comment]
Using 8.3 filenames might help, but a user can disable it. And probably Microsoft will eventually do so too.

**Comment by nikaotix** [comment]
I took some time today to look into this (along with a preliminary std::filesystem implementation for FSNode), and it does seem to me like adding Unicode (at least UTF-8) support internally is a good idea. It seems good enough for now to have a manual conversion from UTF-8 to ASCII, where any non-ASCII character gets turned into '_' or some other character, and use that for the internal font display, but keep all the file/OS interactions in UTF-8 or whatever native encoding is used (AFAIK Windows uses wide characters/UTF-16 for some things). Eventually, the font/text drawing system could be updated for modern fonts, but that's not something that's really needed yet. The only problem with this scheme would be that it may be difficult/impossible to actually save files with filenames that contain the non-ASCII characters, so there may need to be some adjustments if Stella relies on same name but different extension files.

**Comment by sa666666** [comment]
Thanks for looking into this.  I've already done an internal/private implementation using `std::filesystem`, and it ended up being slower than the current platform-specific code.  Since it didn't really solve any other problems and was slower, I decided to stick with what we have now.

But getting the Unicode characters working in Stella would be the main goal here.  Your idea for ASCII <-> Unicode conversion seems plausible.

**Comment by nikaotix** [comment]
I've got a mostly working (ZIP FSNodes aren't working) implementation for Unicode file paths on Windows at https://github.com/nikaotix/stella/tree/origin/feature/windows-unicode. I'll probably take a look at the ZIP stuff later this week. Alternatively, would it make sense to use some of the code from ScummVM for this? Only issue there would be GPL2 vs GPL3. If that's a deal-breaker, no worries.

**Comment by sa666666** [comment]
I just looked over your code, and noticed that your tree has reference to `utf8cpp`, but the files aren't anywhere in the tree.  I googled this, and assume it refers to https://github.com/nemtrif/utfcpp.

Maybe it's better to convert all the file-handling stuff to use UTF correctly, and not just replace certain chars with '_'.  I guess my main concern is whether the following will work.  Note that this is from https://github.com/stella-emu/stella/issues/1084, which is a related issue.  Please take a look at that issue for more info.

A user with username `Usuário` posted that the settings are not properly loading for them.  This is because the username has a Unicode character which Stella is misinterpreting.  It works in Linux and I believe macOS (but I haven't tested the latter).  It also works for me in Windows, but not fin Windows for other members of the Stella team.  The problem with just replacing with '_' is that eventually the settings need to be saved back to the filesystem, so it needs to have the original accented characters present, not the '_'.

Note that I haven't had a chance to test your code yet.  Perhaps this is already addressed.

**Comment by sa666666** [comment]
Ideally I would want to keep the abstraction clean, and at the level of `FSNode` (what the core code interfaces with), this would "just work".  Maybe Windows does something different behind the scenes, but it would be hidden at the `FSNode` abstraction.

If that means we need to change the `FSNode` API itself to use wide strings by default, then I would be fine with that.  What I wouldn't want is in the core code to be doing different things based on whether Windows is being used.  So no `#ifdefs` or anything like that.  And if that means changing `string_view` in path handling everywhere, then so be it.

**Comment by sa666666** [comment]
Maybe the methods for path stuff that take `string_view` should take a `FSNode&` parameter instead, and work directly on the `FSNode`.  That way the methods in `FSNode` can be used directly, and will already be abstracted by the inheritance at `FSNode`.

Note that I'm not implying _you_ have to do this.  Just throwing ideas out there as to how to do this with as little refactoring as possible.

**Comment by nikaotix** [comment]
Just been busy with other things, I’ll see if I can get something that uses
the FileNode pointers instead of paths within a few days.

**Comment by sa666666** [comment]
OK, no problem.

**Comment by sa666666** [comment]
This is now fixed, starting from commit https://github.com/stella-emu/stella/commit/2d5992c2ada7896765b7f6e9f23034b27eacba3a.
